mirrord/intproxy/src/proxies/incoming/interceptor.rs

```diff
@@ -0,0 +1,557 @@
+//! [`BackgroundTask`] used by [`Incoming`](super::IncomingProxy) to manage a single
+//! intercepted connection.
+
+use std::{
+    io::{self, ErrorKind},
+    net::SocketAddr,
+    time::Duration,
+};
+
+use hyper::{StatusCode, Version};
+use mirrord_protocol::tcp::{
+    HttpRequestFallback, HttpResponse, HttpResponseFallback, InternalHttpBody,
+};
+use thiserror::Error;
+use tokio::{
+    io::{AsyncReadExt, AsyncWriteExt},
+    net::{TcpSocket, TcpStream},
+    sync::oneshot::Receiver,
+    time,
+};
+
+use super::http::{HttpSender, UpgradedConnection};
+use crate::background_tasks::{BackgroundTask, MessageBus};
+
+/// Messages consumed by the [`Interceptor`] when it runs as a [`BackgroundTask`].
+pub enum MessageIn {
+    /// Request to be sent to the user application.
+    Http(HttpRequestFallback),
+    /// Data to be sent to the user application.
+    Raw(Vec<u8>),
+}
+
+/// Messages produced by the [`Interceptor`] when it runs as a [`BackgroundTask`].
+#[derive(Debug)]
+pub enum MessageOut {
+    /// Response received from the user application.
+    Http(HttpResponseFallback),
+    /// Data received from the user application.
+    Bytes(Vec<u8>),
+}
+
+impl From<HttpRequestFallback> for MessageIn {
+    fn from(value: HttpRequestFallback) -> Self {
+        Self::Http(value)
+    }
+}
+
+impl From<Vec<u8>> for MessageIn {
+    fn from(value: Vec<u8>) -> Self {
+        Self::Raw(value)
+    }
+}
+
+/// Errors that can occur when [`Interceptor`] runs as a [`BackgroundTask`].
+#[derive(Error, Debug)]
+pub enum Error {
```

Comment:
I'm not a big fan of using rust names like this (same for how we do `type Result = ...` in some places).

I would suggest changing it to `InterceptorError`, but I don't think we have a style rule around this yet, so it's up to you.

Reply:
Changes both to `InterceptorError`/`InterceptorResult` :+1: